sdks/java/io/iceberg/src/main/java/org/apache/beam/io/iceberg/IcebergWriteSchemaTransformProvider.java

```diff
@@ -0,0 +1,216 @@
+
+    public abstract @Nullable String getCatalogImplementation();
+
+    public abstract @Nullable String getWarehouseLocation();
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      public abstract Builder setCatalogName(String catalogName);
+
+      public abstract Builder setCatalogType(String catalogType);
+
+      public abstract Builder setCatalogImplementation(String catalogImplementation);
+
+      public abstract Builder setWarehouseLocation(String warehouseLocation);
+
+      public abstract CatalogConfig build();
+    }
+
+    Set<String> validTypes =
+        Sets.newHashSet(
+            CatalogUtil.ICEBERG_CATALOG_TYPE_HADOOP,
+            CatalogUtil.ICEBERG_CATALOG_TYPE_HIVE,
+            CatalogUtil.ICEBERG_CATALOG_TYPE_REST);
+
+    public void validate() {
+      if (Strings.isNullOrEmpty(getCatalogType())) {
+        checkArgument(
+            validTypes.contains(Preconditions.checkArgumentNotNull(getCatalogType())),
+            "Invalid catalog type. Please pick one of %s",
+            validTypes);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  static class IcebergWriteSchemaTransform extends SchemaTransform {
+    private final Config configuration;
+
+    IcebergWriteSchemaTransform(Config configuration) {
+      this.configuration = configuration;
+    }
+
+    @Override
+    public PCollectionRowTuple expand(PCollectionRowTuple input) {
+
+      PCollection<Row> rows = input.get(INPUT_TAG);
+
+      CatalogConfig catalogConfig = configuration.getCatalogConfig();
+
+      IcebergCatalogConfig.Builder catalogBuilder =
+          IcebergCatalogConfig.builder()
+              .setName(catalogConfig.getCatalogName())
+              .setIcebergCatalogType(catalogConfig.getCatalogType())
+              .setWarehouseLocation(catalogConfig.getWarehouseLocation());
+
+      if (!Strings.isNullOrEmpty(catalogConfig.getCatalogType())) {
+        catalogBuilder = catalogBuilder.setIcebergCatalogType(catalogConfig.getCatalogType());
+      }
+      if (!Strings.isNullOrEmpty(catalogConfig.getWarehouseLocation())) {
+        catalogBuilder = catalogBuilder.setWarehouseLocation(catalogConfig.getWarehouseLocation());
+      }
+
+      // TODO: support dynamic destinations
+      DynamicDestinations dynamicDestinations =
+          DynamicDestinations.singleTable(TableIdentifier.parse(configuration.getTable()));
+
+      IcebergWriteResult result =
+          rows.apply(
+              IcebergIO.writeToDynamicDestinations(catalogBuilder.build(), dynamicDestinations));
+
+      PCollection<Row> snapshots =
+          result
+              .getSnapshots()
+              .apply(MapElements.via(new SnapshotToRow()))
+              .setRowSchema(SnapshotToRow.SNAPSHOT_SCHEMA);
+
+      return PCollectionRowTuple.of(OUTPUT_TAG, snapshots);
+    }
+
+    @VisibleForTesting
+    static class SnapshotToRow extends SimpleFunction<KV<String, Snapshot>, Row> {
+      static final Schema SNAPSHOT_SCHEMA =
```

Comment:
We should document this at the top level.

Reply:
Done